grove/alpha/bernstein_vazirani/bernstein_vazirani.py

```diff
@@ -1,171 +1,203 @@
 """
 
 import numpy as np
 import pyquil.quil as pq
-from pyquil.gates import *
-from six.moves import input
+from pyquil.gates import H, X
+from grove.alpha.bernstein_vazirani import utils
+from collections import defaultdict
 
 
-def oracle_function(vec_a, b, qubits, ancilla):
+def create_bv_bitmap(a, b):
     """
-    Creates a black box oracle for a function
-    to be used in the Bernstein-Vazirani algorithm.
-
-    For a function :math:`f` such that
-
-    .. math::
-
-       f:\\{0,1\\}^n\\rightarrow \\{0,1\\}
-
-       \\mathbf{x}\\rightarrow \\mathbf{a}\\cdot\\mathbf{x}+b\\pmod{2}
+    This function creates a map from bitstring to function value for a boolean formula :math:`f`
+    with
 
-       (\\mathbf{a}\\in\\{0,1\\}^n, b\\in\\{0,1\\})
+        .. math::
 
-    where :math:`(\\cdot)` is the bitwise dot product,
-    this function defines a program that performs
-    the following unitary transformation:
+           f:\\{0,1\\}^n\\rightarrow \\{0,1\\}
 
-    .. math::
+           \\mathbf{x}\\rightarrow \\mathbf{a}\\cdot\\mathbf{x}+b\\pmod{2}
 
-        \\vert \\mathbf{x}\\rangle\\vert y\\rangle \\rightarrow
-        \\vert \\mathbf{x}\\rangle
-        \\vert f(\\mathbf{x}) \\text{ xor } y\\rangle
-
-    where :math:`\\text{xor}` is taken bitwise.
-
-    Allocates one scratch bit.
-
-    :param list(int) vec_a: a vector of length :math:`n`
-                                containing only ones and zeros.
-                                The order is taken to be
-                                most to least significant bit.
-    :param int b: a 0 or 1
-    :param list(int) qubits: List of qubits that enter as input
-                             :math:`\\vert x\\rangle`.
-                             Must be the same length (:math:`n`)
-                             as :math:`\\mathbf{a}`
-    :param int ancilla: Ancillary qubit to serve as input
-                        :math:`\\vert y\\rangle`,
-                        where the answer will be written to.
-    :return: A program that performs the above unitary transformation.
-    :rtype: Program
-    """
-    assert len(vec_a) == len(qubits), \
-        "vec_a must be the same length as the number of input qubits"
-    assert all(list(map(lambda x: x in {0, 1}, vec_a))), \
-        "vec_a must be a list of 0s and 1s"
-    assert b in {0, 1}, "b must be a 0 or 1"
-
-    n = len(qubits)
-    p = pq.Program()
-    if b == 1:
-        p.inst(X(ancilla))
-    for i in range(n):
-        if vec_a[i] == 1:
-            p.inst(CNOT(qubits[n - 1 - i], ancilla))
-    return p
-
-
-def bernstein_vazirani(oracle, qubits, ancilla):
+           (\\mathbf{a}\\in\\{0,1\\}^n, b\\in\\{0,1\\})
+    :param String a: a string of 0's and 1's that represents the dot-product partner in :math:`f`
+    :param String b: 0 or 1 as a string representing the bias term in :math:`f`
```

Comment:
The listed reference at the top doesn't talk about a bias term.

Reply:
Ah, but the docs do.

Reply:
👍